experimental/virtio/src/vsock/mod.rs

```diff
@@ -14,4 +14,178 @@
 // limitations under the License.
 //
 
+use crate::queue::{DeviceWriteOnlyQueue, DriverWriteOnlyQueue};
+use anyhow::Context;
+use packet::Packet;
+use rust_hypervisor_firmware_subset::{
+    device::VirtioBaseDevice,
+    pci::{find_device, VirtioPciTransport},
+};
+
 pub mod packet;
+
+/// The number of buffer descriptors in each of the queues.
+const QUEUE_SIZE: usize = 16;
+
+/// The size of each of the buffers used by the transmit and receive queues.
+const DATA_BUFFER_SIZE: usize = 4096;
+
+/// The size of the buffers used by the event queue.
+const EVENT_BUFFER_SIZE: usize = 8;
+
+/// The id of the receive queue.
+const RX_QUEUE_ID: u16 = 0;
+
+/// The id of the transmit queue.
+const TX_QUEUE_ID: u16 = 1;
+
+/// The id of the event queue.
+const EVENT_QUEUE_ID: u16 = 2;
+
+/// The vendor ID for virtio PCI devices.
+const PCI_VENDOR_ID: u16 = 0x1AF4;
+
+/// The virtio device id for a socket device.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-1930005>.
+const DEVICE_ID: u16 = 19;
+
+/// The PCI device id for a virtio socket device.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-1020002>.
+const PCI_DEVICE_ID: u16 = 0x1040 + DEVICE_ID;
+
+/// The PCI config offset for reading the guest CID.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-3940004>.
+const CID_CONFIG_OFFSET: u64 = 0;
+
+/// The well-known host CID value.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-3940004>.
+const HOST_CID: u64 = 2;
+
+/// Low-level driver interface to interact with a virtio socket device.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-39000010>.
+pub struct VSock {
+    /// The base virtio-over-PCI device used for configuration and notification.
+    device: VirtioBaseDevice<VirtioPciTransport>,
+    /// The transmit queue.
+    tx_queue: DriverWriteOnlyQueue<QUEUE_SIZE, DATA_BUFFER_SIZE>,
+    /// The receive queue.
+    rx_queue: DeviceWriteOnlyQueue<QUEUE_SIZE, DATA_BUFFER_SIZE>,
+    /// The event queue used by the device to notify the driver that the guest CID has changed. We
+    /// ignore it for now as we don't support live migration, but it still must exist and be
+    /// configured.
+    event_queue: DeviceWriteOnlyQueue<QUEUE_SIZE, EVENT_BUFFER_SIZE>,
+    /// The the CID assigned to this VM.
+    guest_cid: u64,
+    /// Whether the device have been initialized and is ready to use.
+    ready: bool,
```

Comment:
Is it ever possible to have a device that is not ready? Looks like `find_and_configure_device()` calls `init()`, so it's impossible to have it not initialized.